lambdas/src/apis/collections/controllers/collections.ts

```diff
@@ -150,46 +200,96 @@ function buildUrn(protocol: string, contract: string, option: string): string {
+  if (!isValidSize(size)) throw new ServiceError('Invalid size')
+  return true
+}
+
+function getFileHash(entity: Entity, fileName?: string): string {
+  const hash = findHashForFile(entity, fileName)
+  if (!hash) throw new ServiceError(`Hash not found for file ${fileName}`, 404)
+
+  return hash
 }
 
-async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity | undefined> {
+async function fetchEntity(client: SmartContentClient, urn: string): Promise<Entity> {
   const entities: Entity[] = await client.fetchEntitiesByPointers(EntityType.WEARABLE, [urn])
-  return entities && entities.length > 0 && entities[0].metadata ? entities[0] : undefined
+  if (!(entities && entities.length > 0 && entities[0].metadata)) throw new ServiceError('Entity not found', 404)
+
+  return entities[0]
 }
 
-export function getBodyShapes(representations: WearableMetadataRepresentation[]) {
+function getBodyShapes(representations: WearableMetadataRepresentation[]): string[] {
   const bodyShapes = new Set<WearableBodyShape>()
   for (const representation of representations) {
     for (const bodyShape of representation.bodyShapes) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       bodyShapes.add(bodyShape.split(':').pop()!)
     }
   }
   return Array.from(bodyShapes)
 }
 
+function isValidSize(size: string): size is ValidSize {
+  return sizes[size] !== undefined
+}
+
+function getImagePath(root: string, imageRequest: ImageRequest): string {
+  return root + `/` + getImageRequestId(imageRequest) + '.png'
+}
+
+function getImageRequestId({ urn, hash, size, rarity }: ImageRequest): string {
+  return `${urn}-${hash}-${size}` + (rarity ? `-${rarity}` : '')
+}
+
+async function getImage(
+  client: SmartContentClient,
+  rootStorageLocation: string,
+  imageRequest: ImageRequest
+): Promise<Buffer> {
+  const path = getImagePath(rootStorageLocation, imageRequest)
+
+  // Check if the image is already in the cache, otherwise build and store it
+  fs.existsSync(path) || (await buildImage(client, rootStorageLocation, imageRequest))
+
+  return await sharp(path).toBuffer()
+}
+
+async function buildImage(client: SmartContentClient, rootStorageLocation: string, imageRequest: ImageRequest) {
+  const imagePath = getImagePath(rootStorageLocation, imageRequest)
+  fs.existsSync(imagePath) || (await saveImage(client, rootStorageLocation, imageRequest))
+}
+
+async function saveImage(client: SmartContentClient, rootStorageLocation: string, imageRequest: ImageRequest) {
+  const image = await client.downloadContent(imageRequest.hash)
+  const imagePath = getImagePath(rootStorageLocation, imageRequest)
+  let finalImage: sharp.Sharp
+  if (imageRequest.rarity) {
+    const background = await getRarityBackground(imageRequest.rarity)
+    const shouldResize = sizes[imageRequest.size] !== DEFAULT_IMAGE_SIZE
+    const resizedBackground = shouldResize
```

Comment:
Should we store the resized backgrounds too or is it not necessary?

Reply:
yep, it can be a good idea